File: lab1/crypto/crypto.cpp
#include <iostream>
#include <fstream>
#include <cryptopp/aes.h>
#include <cryptopp/rsa.h>
#include <cryptopp/osrng.h>
#include <cryptopp/files.h>
#include <cryptopp/modes.h>

int main() {
    CryptoPP::AutoSeededRandomPool rng;

    // Генерация закрытого ключа RSA (2048 бит)
    CryptoPP::RSA::PrivateKey privateKey;
    privateKey.GenerateRandomWithKeySize(rng, 2048);

    // Сохранение закрытого ключа в файл
    CryptoPP::FileSink privateKeyFile("private_key.pem");
    privateKey.Save(privateKeyFile);

    std::cout << "Закрытый ключ RSA сгенерирован." << std::endl;

    // Получение публичного ключа из закрытого ключа
    CryptoPP::RSA::PublicKey publicKey;
    publicKey = privateKey;

    // Сохранение публичного ключа в файл
    CryptoPP::FileSink publicKeyFile("public_key.pem");
    publicKey.Save(publicKeyFile);

    std::cout << "Публичный ключ RSA извлечен." << std::endl;

    // Генерация случайного симметричного ключа AES
    CryptoPP::SecByteBlock symmetricKey(32); // Выберите длину ключа AES
    rng.GenerateBlock(symmetricKey, symmetricKey.size());

    // Сохранение симметричного ключа в файл
    std::ofstream symmetricKeyFile("symmetric_key.bin", std::ios::binary);
    symmetricKeyFile.write(reinterpret_cast<char*>(symmetricKey.BytePtr()), symmetricKey.size());
    symmetricKeyFile.close();

    std::cout << "Симметричный ключ AES сгенерирован." << std::endl;

    // Загрузка симметричного ключа из файла
    std::ifstream symmetricKeyFileIn("symmetric_key.bin", std::ios::binary);
    symmetricKeyFileIn.read(reinterpret_cast<char*>(symmetricKey.BytePtr()), symmetricKey.size());
    symmetricKeyFileIn.close();

    // Шифрование данных с использованием симметричного ключа (AES)
    std::string plaintext = "Oleksii";
    std::string ciphertext;
    
    CryptoPP::byte iv[CryptoPP::AES::BLOCKSIZE];
    rng.GenerateBlock(iv, sizeof(iv));


    CryptoPP::AES::Encryption aesEncryption(symmetricKey.BytePtr(), symmetricKey.size());
    CryptoPP::CBC_Mode_ExternalCipher::Encryption cbcEncryption(aesEncryption, iv);

    CryptoPP::StreamTransformationFilter encryptor(cbcEncryption, new CryptoPP::StringSink(ciphertext));
    encryptor.Put(reinterpret_cast<const unsigned char*>(plaintext.data()), plaintext.size());
    encryptor.MessageEnd();

    std::cout << "Данные успешно зашифрованы с использованием симметричного ключа AES." << std::endl;

    // Шифрование симметричного ключа с использованием публичного ключа RSA
    CryptoPP::RSAES_OAEP_SHA_Encryptor rsaEncryptor(publicKey);
    std::string encryptedSymmetricKey;

    CryptoPP::StringSource s(symmetricKey.BytePtr(), symmetricKey.size(), true, new CryptoPP::PK_EncryptorFilter(rng, rsaEncryptor, new CryptoPP::StringSink(encryptedSymmetricKey)));

    std::cout << "Симметричный ключ успешно зашифрован с использованием публичного ключа RSA." << std::endl;

    // Расшифрование симметричного ключа с использованием закрытого ключа RSA
    CryptoPP::RSAES_OAEP_SHA_Decryptor rsaDecryptor(privateKey);
    std::string decryptedSymmetricKey;

    CryptoPP::StringSource ss(encryptedSymmetricKey, true, new CryptoPP::PK_DecryptorFilter(rng, rsaDecryptor, new CryptoPP::StringSink(decryptedSymmetricKey)));

    std::cout << "Симметричный ключ успешно расшифрован с использованием закрытого ключа RSA." << std::endl;

    // Расшифрование данных симметричным ключом (AES)
    std::string recoveredPlaintext;

    CryptoPP::AES::Decryption aesDecryption(symmetricKey.BytePtr(), symmetricKey.size());
    CryptoPP::CBC_Mode_ExternalCipher::Decryption cbcDecryption(aesDecryption, iv);

    CryptoPP::StreamTransformationFilter decryptor(cbcDecryption, new CryptoPP::StringSink(recoveredPlaintext));
    decryptor.Put(reinterpret_cast<const unsigned char*>(ciphertext.data()), ciphertext.size());
    decryptor.MessageEnd();

    std::cout << "Данные успешно расшифрованы с использованием симметричного ключа AES." << std::endl;

    // Вывод результатов
    std::cout << "Зашифрованный ключ RSA: " << encryptedSymmetricKey << std::endl;
    std::cout << "Расшифрованный ключ AES: " << decryptedSymmetricKey << std::endl;
    std::cout << "Зашифрованный текст: " << ciphertext << std::endl;
    std::cout << "Расшифрованный текст: " << recoveredPlaintext << std::endl;

    return 0;
}

File: lab1/Chalyi_FB-21mn/pycrypto/crypto.py
#!/usr/bin/env python

from Crypto.Cipher import AES, PKCS1_OAEP
from Crypto.PublicKey import RSA
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad

# Генерация закрытого ключа RSA (2048 бит)
private_key = RSA.generate(2048)

# Сохранение закрытого ключа в файл
private_key_file = open("private_key.pem", "wb")
private_key_file.write(private_key.export_key())
private_key_file.close()

print("Закрытый ключ RSA сгенерирован.")

# Получение публичного ключа из закрытого ключа
public_key = private_key.publickey()

# Сохранение публичного ключа в файл
public_key_file = open("public_key.pem", "wb")
public_key_file.write(public_key.export_key())
public_key_file.close()

print("Публичный ключ RSA извлечен.")

# Генерация случайного симметричного ключа AES
symmetric_key = get_random_bytes(16)  # 16 байт для AES-128

# Сохранение симметричного ключа в файл
symmetric_key_file = open("symmetric_key.bin", "wb")
symmetric_key_file.write(symmetric_key)
symmetric_key_file.close()

print("Симметричный ключ AES сгенерирован.")

# Загрузка симметричного ключа из файла
symmetric_key_file_in = open("symmetric_key.bin", "rb")
symmetric_key = symmetric_key_file_in.read()
symmetric_key_file_in.close()

# Шифрование данных с использованием симметричного ключа (AES)
plaintext = "Oleksii"
iv = get_random_bytes(16)  # Генерация случайного вектора инициализации
cipher = AES.new(symmetric_key, AES.MODE_CBC, iv)

# Добавляем дополнение PKCS7
padded_plaintext = pad(plaintext.encode('utf-8'), AES.block_size)
ciphertext = iv + cipher.encrypt(padded_plaintext)

# Шифрование симметричного ключа с использованием публичного ключа RSA
rsa_cipher = PKCS1_OAEP.new(public_key)
encrypted_symmetric_key = rsa_cipher.encrypt(symmetric_key)

encrypted_symmetric_key_file = open("encrypted_symmetric_key.bin", "wb")
encrypted_symmetric_key_file.write(encrypted_symmetric_key)
encrypted_symmetric_key_file.close()
print("Симметричный ключ успешно зашифрован с использованием публичного ключа RSA.")

# Расшифрование симметричного ключа с использованием закрытого ключа RSA
rsa_decipher = PKCS1_OAEP.new(private_key)
decrypted_symmetric_key = rsa_decipher.decrypt(encrypted_symmetric_key)

decrypted_symmetric_key_file = open("decrypted_symmetric_key.bin", "wb")
decrypted_symmetric_key_file.write(decrypted_symmetric_key)
decrypted_symmetric_key_file.close()
print("Симметричный ключ успешно расшифрован с использованием закрытого ключа RSA.")

# Расшифрование данных симметричным ключом (AES)
iv = ciphertext[:16]
cipher = AES.new(decrypted_symmetric_key, AES.MODE_CBC, iv)
padded_plaintext = cipher.decrypt(ciphertext[16:])
plaintext = unpad(padded_plaintext, AES.block_size).decode('utf-8')

# Вывод зашифрованного ключа RSA
print("Зашифрованный ключ RSA:", encrypted_symmetric_key.hex())

# Вывод расшифрованного ключа AES
print("Расшифрованный ключ AES:", decrypted_symmetric_key.hex())

# Вывод зашифрованного текста
print("Зашифрованный текст:", ciphertext.hex())
print("Данные успешно зашифрованы с использованием симметричного ключа AES.")

print("Данные успешно расшифрованы с использованием симметричного ключа AES.")
print("Расшифрованный текст:", plaintext)
